src/python/pants/backend/visibility/rule_types.py

```diff
@@ -84,3 +101,152 @@ def parse(cls, arg: Any) -> VisibilityRuleSet:
+                f"There is no matching dependencies rule for the `{source_adaptor.type_alias}` "
+                f"target {source_path}:{source_adaptor.name or os.path.basename(source_path)} "
+                f"for the dependency on the `{target_adaptor.type_alias}` target {target_path}:"
+                f"{target_adaptor.name or os.path.basename(target_path)} in {dep_rules.path}"
+            )
+        if outgoing == DependencyRuleAction.DENY:
+            return outgoing
+
+        # Check incoming dependency action
+        incoming = (
+            cast(BuildFileVisibilityRules, dependents_rules).get_action(
+                target_adaptor,
+                source_path,
+                relpath=target_path,
+                outgoing_dependency=False,
+            )
+            if dependents_rules is not None
+            else DependencyRuleAction.ALLOW
+        )
+        if incoming is None:
+            dep_rules = cast(BuildFileVisibilityRules, dependents_rules)
+            raise BuildFileVisibilityRulesError(
+                f"There is no matching dependents rule for the `{target_adaptor.type_alias}` "
+                f"target {target_path}:{target_adaptor.name or os.path.basename(target_path)} "
+                f"for the dependency from the `{source_adaptor.type_alias}` target {source_path}:"
+                f"{source_adaptor.name or os.path.basename(source_path)} in {dep_rules.path}"
+            )
+        return incoming if incoming != DependencyRuleAction.ALLOW else outgoing
+
+    def get_action(
+        self,
+        target: TargetAdaptor,
+        path: str,
+        relpath: str,
+        outgoing_dependency: bool,
+    ) -> DependencyRuleAction | None:
+        """Get applicable rule for target type from `path`.
+
+        The rules are declared in `relpath`.
+        """
+        ruleset = self.get_ruleset(target)
+        if ruleset is None:
+            return None
+        for visibility_rule in ruleset.rules:
+            if visibility_rule.match(path, relpath):
+                if visibility_rule.action != DependencyRuleAction.ALLOW:
+                    target_addr = f"{relpath}:{target.name or os.path.basename(relpath)}"
+                    target_type = target.type_alias
+                    if outgoing_dependency:
+                        logger.debug(
+                            f"{visibility_rule.action.name}: the `{target_type}` target "
+                            f"{target_addr} dependency to a target at {path} by dependency rule "
+                            f"{visibility_rule} declared in {self.path}"
+                        )
+                    else:
+                        logger.debug(
+                            f"{visibility_rule.action.name}: dependency on the `{target_type}` "
+                            f"target {target_addr} from a target at {path} by dependent rule "
+                            f"{visibility_rule} declared in {self.path}"
```

Comment:
these log messages will be crucial to understand why a certain dependency was denied in case of non-trivial setups.

I did consider if we should include the source of the rule with the returned rule action to offer greater flexibility in when where and how this is presented.